grpc/src/main/java/com/linecorp/armeria/server/grpc/DeferredListener.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.grpc;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+import io.netty.util.concurrent.EventExecutor;
+
+final class DeferredListener<I> extends ServerCall.Listener<I> {
+
+    private static final List<?> NOOP_TASKS = ImmutableList.of();
+
+    @Nullable
+    private final Executor blockingExecutor;
+    @Nullable
+    private final EventExecutor eventLoop;
+
+    private List<Consumer<Listener<I>>> pendingTasks = new ArrayList<>();
+    @Nullable
+    private Listener<I> delegate;
+
+    private volatile boolean callClosed;
```

Comment:
Is this field accessed by two threads?

Reply:
`callClosed` is written by `sequentialExecutor()` and is accessed by any threads.
It would be safe to remove `volatile` because it is doubled checked when a pending task is running in `sequentialExecutor().execute()`

Reply:
I don't get it. 😅 It is accessed by any threads and it would be safe to remove `volatile`?

Reply:
If `callClosed` is false in a non-sequential executor, it is re-read when it is submitted to the sequential executor. It's likely that it would be better to remove `volatile` consistency.
Most gRPC services use unary calls and it starts a call when a request is fully aggregated. So the callbacks may be called synchronously before the async task is completed.

Reply:
Now I get it. Thanks for the explanation. 👍